src/plugin/nerd_plugin.js

```diff
@@ -166,122 +218,141 @@ class NerdPluginVirtual extends EventEmitter {
+    yield this.rpc.call('send', [transfer, settleAddress])
+
+    if (!transfer.executionCondition) {
+      yield this.balance.add(transfer.amount)
+      this.emit('outgoing_transfer', transfer)
+    } else {
+      this.emit('outgoing_prepare', transfer)
+    }
+
+    yield this._completeTransfer(transfer)
+    this._handleTimer(transfer)
+
+    return Promise.resolve(null)
   }
 
   getInfo () {
     return Promise.resolve(this.info)
   }
 
   getFulfillment (transferId) {
-    return this._getFulfillment(transferId).then((fulfillment) => {
-      if (!fulfillment) {
-        return Promise.reject(null)
-      } else {
-        return Promise.resolve(fulfillment)
-      }
-    })
+    return co.wrap(this._getFulfillment).call(this, transferId)
   }
 
-  _getFulfillment (transferId) {
-    return this.transferLog.getFulfillment(transferId)
+  * _getFulfillment (transferId) {
+    const stored = yield this.transferLog.get(transferId)
+    if (!stored) {
+      throw new TransferNotFoundError(transferId + ' not found')
+    }
+
+    const fulfillment = yield this.transferLog.getFulfillment(transferId)
+    if (!fulfillment) {
+      throw new MissingFulfillmentError(transferId + ' has not been fulfilled')
+    }
+
+    return fulfillment
   }
 
   fulfillCondition (transferId, fulfillmentBuffer) {
     let fulfillment = fulfillmentBuffer.toString()
-    let transfer = null
     this._log('fulfilling: ' + fulfillment)
-    return this.transferLog.get(transferId).then((storedTransfer) => {
-      transfer = storedTransfer
-      return this._fulfillConditionLocal(transfer, fulfillment)
-    }).catch(this._handle)
+
+    return co.wrap(this._fulfillConditionLocal).call(this, transferId, fulfillment)
+      .then((res) => {
+        this.rpc.call('fulfill', [res.transfer, res.direction])
+        return Promise.resolve(null)
+      })
   }
 
-  rejectIncomingTransfer (transferId) {
-    let transfer = null
+  * _fulfillConditionLocal (transferId, fulfillment) {
+    const transfer = yield this.transferLog.get(transferId)
+
+    if (!transfer || !transfer.executionCondition) {
+      throw new TransferNotFoundError('no conditional transfer exists with the given id')
+    }
+
+    const fulfilled = yield this.transferLog.isFulfilled(transfer.id)
+    if (fulfilled) {
+      throw new RepeatError('this transfer has already been fulfilled')
+    }
+
+    try {
+      cc.fromFulfillmentUri(fulfillment)
+    } catch (e) {
+      throw new InvalidFieldsError('malformed fulfillment')
+    }
+
+    const execute = transfer.executionCondition
+    const time = new Date()
+    const expiresAt = new Date(transfer.expiresAt)
+    const timedOut = (time > expiresAt)
+
+    if (this._validate(fulfillment, execute) && !timedOut) {
+      return yield this._executeTransfer(transfer, fulfillment)
+    } else if (timedOut) {
+      yield this._timeOutTransfer(transfer)
+      throw new RepeatError('this transfer has already timed out')
```

Comment:
Should be a different error type, trying to fulfill an expired transfer is not a repetition.